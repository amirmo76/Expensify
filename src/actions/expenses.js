import uuid from "uuid";

export const addExpense = ({
  description = "",
  note = "",
  amount = 0,
  createdAt = 0
} = {}) => ({
  type: "ADD_EXPENSE",
  expense: {
    id: uuid(),
    description,
    note,
    amount,
    createdAt
  }
});

export const removeExpense = ({ id } = {}) => {
  if (!id) {
    throw new Error("not specifying the expense id to remove!");
  }

  return {
    type: "REMOVE_EXPENSE",
    id
  };
};

export const editExpense = ({ id, data } = {}) => {
  if (!id) {
    throw new Error("not specifying the expense id to edit!");
  }

  return {
    type: "EDIT_EXPENSE",
    id,
    data
  };
};
