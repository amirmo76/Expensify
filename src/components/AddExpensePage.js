import React from "react";

const AddExpensePage = () => {
  return <div>This is from Add Expense Page</div>;
};

export default AddExpensePage;
