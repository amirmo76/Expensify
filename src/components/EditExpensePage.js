import React from "react";

const EditExpensePage = () => {
  return <div>This is from Edit Expense Page</div>;
};

export default EditExpensePage;
