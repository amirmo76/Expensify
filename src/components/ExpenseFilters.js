import React from "react";
import { connect } from "react-redux";
import {
  editTextFilter,
  sortByAmount,
  sortByDate,
  setEndDate,
  setStartDate
} from "../actions/filters";
import { DateRangePicker } from "react-dates";

class ExpenseFilters extends React.Component {
  state = {
    dateRangePickerFocused: null
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.dispatch(setStartDate(startDate));
    this.props.dispatch(setEndDate(endDate));
  };

  render() {
    return (
      <div>
        <h2>Filters</h2>
        <input
          value={this.props.filters.text}
          onChange={e => {
            this.props.dispatch(editTextFilter({ text: e.target.value }));
          }}
        />
        <select
          value={this.props.filters.sortBy}
          onChange={e => {
            switch (e.target.value) {
              case "amount":
                this.props.dispatch(sortByAmount());
                break;

              case "date":
                this.props.dispatch(sortByDate());
                break;
            }
          }}
        >
          <option value="date" defaultValue>
            Date
          </option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          numberOfMonths={1}
          isOutsideRange={() => false}
          focusedInput={this.state.dateRangePickerFocused}
          onFocusChange={focused =>
            this.setState(() => ({ dateRangePickerFocused: focused }))
          }
          onDatesChange={this.onDatesChange}
          startDateId="date_range_start"
          endDateId="date_range_end"
          showClearDates={true}
        />
      </div>
    );
  }
}

const mapStateToProps = state => ({
  filters: state.filters
});

export default connect(mapStateToProps)(ExpenseFilters);
