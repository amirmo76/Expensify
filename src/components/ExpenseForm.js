import React, { Component } from 'react';

class ExpenseForm extends Component {

  render() {

  }
} 
