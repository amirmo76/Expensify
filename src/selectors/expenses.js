export const getVisibleExpenses = (
  expenses,
  { text, sortBy, startDate, endDate }
) => {
  return expenses
    .filter(cur => {
      const textMatch = cur.description
        .toLowerCase()
        .includes(text.toLowerCase());

      const startDateMatch =
        typeof startDate !== "number" || cur.createdAt >= startDate;

      const endDateMatch =
        typeof endDate !== "number" || cur.createdAt <= endDate;

      return textMatch && startDateMatch && endDateMatch;
    })
    .sort((a, b) => {
      if (sortBy === "amount") {
        return b.amount - a.amount;
      }
      return b.createdAt - a.createdAt;
    });
};
