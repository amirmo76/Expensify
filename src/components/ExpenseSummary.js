import React from "react";
import { connect } from "react-redux";
import { getVisibleExpenses } from "../selectors/expenses";
import { getExpensesTotal } from "../selectors/expenses-total";
import numeral from "numeral";

export const ExpenseSummary = ({ expensesCount, expensesTotal }) => {
  return (
    <p>{`Viewing ${expensesCount} expense${
      expensesCount === 1 ? "" : "s"
    } totalling ${numeral(expensesTotal / 100).format("$0,0.00")}`}</p>
  );
};

const mapStateToProps = state => {
  const visibleExpenses = getVisibleExpenses(state.expenses, state.filters);
  return {
    expensesTotal: getExpensesTotal(visibleExpenses),
    expensesCount: visibleExpenses.length
  };
};

export default connect(mapStateToProps)(ExpenseSummary);
