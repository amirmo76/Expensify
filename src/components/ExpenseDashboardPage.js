import React from "react";
import ExpenseList from "./ExpenseList";
import ExpenseFilters from "./ExpenseFilters";
import ExpenseSummary from "./ExpenseSummary";

const ExpenseDashboardPage = () => {
  return (
    <div>
      <ExpenseSummary />
      <ExpenseFilters />
      <ExpenseList />
    </div>
  );
};

export default ExpenseDashboardPage;
