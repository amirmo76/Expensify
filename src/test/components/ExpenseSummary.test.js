import React from "react";
import { shallow } from "enzyme";
import { ExpenseSummary } from "../../components/ExpenseSummary";
import expenses from "../fixtures/expenses";

test("should render ExpenseSummary with 1 expense correctly", () => {
  const wrapper = shallow(
    <ExpenseSummary expensesCount={1} expensesTotal={209567} />
  );

  expect(wrapper).toMatchSnapshot();
});

test("should render ExpenseSummary with more than one expense correctly", () => {
  const wrapper = shallow(
    <ExpenseSummary expensesCount={2} expensesTotal={9823210} />
  );

  expect(wrapper).toMatchSnapshot();
});

test("should render ExpenseSummary with no expense correctly", () => {
  const wrapper = shallow(
    <ExpenseSummary expensesCount={0} expensesTotal={0} />
  );

  expect(wrapper).toMatchSnapshot();
});
