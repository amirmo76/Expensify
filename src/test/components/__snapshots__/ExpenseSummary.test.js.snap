// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseSummary with 1 expense correctly 1`] = `
<p>
  Viewing 1 expense totalling $2,095.67
</p>
`;

exports[`should render ExpenseSummary with more than one expense correctly 1`] = `
<p>
  Viewing 2 expenses totalling $98,232.10
</p>
`;

exports[`should render ExpenseSummary with no expense correctly 1`] = `
<p>
  Viewing 0 expenses totalling $0.00
</p>
`;
